develop/tutorials/articles/06-web-applications/05-jsf/01-packaging-a-jsf-application.markdown

```diff
@@ -0,0 +1,161 @@
+`WEB-INF/faces-config.xml` descriptor and other Java EE resources such as the
+`WEB-INF/web.xml` descriptor. In this tutorial, you'll learn two ways to package
+a JSF portlet so it can be treated as an OSGi module at runtime, and the pros
+and cons behind both.
+
+Although the term *WAR* sounds like a non-OSGi style plugin, there are certain
+steps that can be taken to ensure a WAR plugin is still treated as an OSGi
+module during runtime. You can create your portlet as an OSGi Web Application
+Bundle (WAB), which follows a WAR layout and ends with the usual `.war`
+extension. A WAB is an OSGi bundle version of a WAR file. Moreover, it's a
+portlet organized in a WAR-shaped layout that includes OSGi manifest entries.
+
+Liferay 7.0 supports the WAB standard for deployment of Java EE style WARs.
+In detail, Liferay implements the
+[WAB Extender](https://github.com/liferay/liferay-portal/tree/master/modules/apps/foundation/portal-osgi-web/portal-osgi-web-wab-extender),
+which enables WABs to run as OSGi modules at runtime. You must format your WAB
+so that it is recognized by the WAB Extender and properly deployed into the OSGi
+container.
+
+To get an idea of how a JSF WAB is organized, an example JSF WAB portlet anatomy
+structure is outlined below:
+
+- `META-INF`
+    - `MANIFEST.MF` (OSGi-ready)
+- `OSGI-INF`
+    - Declarative Services descriptors (prepared by `bndlib`)
+- `WEB-INF`
+    - `classes`
+        - Class files and related properties
+    - `lib`
+        - JAR dependencies
+    - `resources`
+        - CSS, XHTML, PNG or other frontend files
+    - `views`
+        - XHTML views
+    - `faces-config.xml`
+    - `web.xml`
+
+To ensure your WAR is recognized by the Liferay's WAB extender, you'll need to
+make sure you're including the `META-INF/MANIFEST.MF` file. You can either
+include a `Bundle-SymbolicName` OSGi directive in this file or refrain from
+including it. Both ways work, but there are differing benefits and drawbacks for
+why to include/not include this directive.
+
+## Including the Bundle-SymbolicName OSGi Directive in Your WAB
+
+**Benefits:**
+
+- Can supply a `bnd.bnd` file and utilize a build-time plugin (e.g.,
+  [bnd-maven-plugin](http://njbartlett.name/2015/03/27/announcing-bnd-maven-plugin.html))
+  to affect the content of an OSGi-ready `META-INF/MANIFEST.MF`.
+- Can take advantage of OSGi Declarative Services (DS)
+- Provides a class that extends.
+  [GenericPortlet](http://portals.apache.org/pluto/portlet-2.0-apidocs/javax/portlet/GenericPortlet.html)
+  (or in the case of JSF,
+  [GenericFacesPortlet](http://myfaces.apache.org/portlet-bridge/2.0/api/apidocs/javax/portlet/faces/GenericFacesPortlet.html))
+  and specify portlet attributes via the `@Component` annotation, rather than
+  using two separate descriptors: `WEB-INF/portlet.xml` and
+  `WEB-INF/liferay-portlet.xml`.
+
+**Drawbacks:**
+
+- The `bnd.bnd` file must include embedded dependency information that is
```

Comment:
Today I learned that Vernon is working on a patch to the bnd-maven-plugin that removes this drawback. It will automatically add all dependencies that are not <scope>provided</scope> so that developers do not have to repeat themselves in bnd.bnd.